### ISSUE TITLE
`not_to` not negating `redirect_to` and `render_template`

### ISSUE THREAD

**Issue by mattfreer-mdsol** [rspec/rspec-rails#1539]
Hi I have a some RSpec tests that follow the following format:

```ruby
expect(response).not_to render_template(:index)
expect(response).not_to redirect_to posts_url
```
In both cases despite the subject of the test functioning as expected, the tests fail. The reason looks like it is a result of the assertions not being negated when using `not_to`.

For example when I run:
```ruby
expect(response).not_to redirect_to posts_url
```
I get:
```
Expected response to be a <redirect>, but was <200>.
```

When I run :
```ruby
expect(response).not_to render_template(:index)
```
I get:
```
expecting <"index"> but rendering with <["applications/foobar", "layouts/application"]>
```

I am running Ruby 2.1.7 with the following dependencies:

  * rails (4.0.13)
  * rspec (3.3.0)
  * rspec-collection_matchers (1.1.2)
  * rspec-core (3.3.1)
  * rspec-expectations (3.3.0)
  * rspec-mocks (3.3.1)
  * rspec-rails (3.3.2)
  * rspec-support (3.3.0)

FYI: I'm not using shoulda-matchers, so [this closed issue](https://github.com/rspec/rspec-rails/issues/1233) looks different.

Thanks.

**Comment by JonRowe** [comment]
It's possible those matchers don't negate properly, but I'm surprised we've never received a bug report before if thats the case! In the mean time try this:

```Ruby
RSpec::Matchers.define_negated_matcher :avoid_rendering_template, :render_template
RSpec::Matchers.define_negated_matcher :avoid_redirect_to, :redirect_to

specify { expect(response).to avoid_rendering_template(:index) }
specify { expect(response).to avoid_redirect_to posts_url }
```

**Comment by mattfreer-mdsol** [comment]
Hi @JonRowe,

Thanks for the suggested workaround. However this also gives me unexpected results. 
Having defined the negated matchers as you suggest, I still have failing tests, with similar output:

When I run:
```ruby
expect(response).to avoid_rendering_template(:index)
```

I get:
```ruby
expecting <"index"> but rendering with <["error/errors", "layouts/foobar"]>.
```

Stacktrace:
```bash
#../ruby/2.1.0/gems/test-unit-3.1.2/lib/test/unit/assertions.rb:55:in `block in assert_block'
#../ruby/2.1.0/gems/test-unit-3.1.2/lib/test/unit/assertions.rb:1627:in `call'
#../ruby/2.1.0/gems/test-unit-3.1.2/lib/test/unit/assertions.rb:1627:in `_wrap_assertion'
#../ruby/2.1.0/gems/test-unit-3.1.2/lib/test/unit/assertions.rb:53:in `assert_block'
#../ruby/2.1.0/gems/test-unit-3.1.2/lib/test/unit/assertions.rb:163:in `block in assert'
#../ruby/2.1.0/gems/test-unit-3.1.2/lib/test/unit/assertions.rb:1632:in `call'
#../ruby/2.1.0/gems/test-unit-3.1.2/lib/test/unit/assertions.rb:1632:in `_wrap_assertion'
#../ruby/2.1.0/gems/test-unit-3.1.2/lib/test/unit/assertions.rb:130:in `assert'
#../ruby/2.1.0/gems/rspec-rails-3.3.2/lib/rspec/rails/adapters.rb:210:in `block (2 levels) in define_assertion_delegato
rs'
#../ruby/2.1.0/gems/actionpack-4.0.13/lib/action_controller/test_case.rb:118:in `assert_template'
#../ruby/2.1.0/gems/rspec-rails-3.3.2/lib/rspec/rails/matchers/have_rendered.rb:17:in `block in matches?'
#../ruby/2.1.0/gems/rspec-expectations-3.3.0/lib/rspec/matchers/built_in/base_matcher.rb:47:in `match_unless_raises'
#../ruby/2.1.0/gems/rspec-rails-3.3.2/lib/rspec/rails/matchers/have_rendered.rb:16:in `matches?'
#../ruby/2.1.0/gems/rspec-expectations-3.3.0/lib/rspec/matchers/matcher_delegator.rb:17:in `method_missing'
#../ruby/2.1.0/gems/rspec-expectations-3.3.0/lib/rspec/matchers/aliased_matcher.rb:25:in `method_missing'
#../ruby/2.1.0/gems/rspec-expectations-3.3.0/lib/rspec/matchers/aliased_matcher.rb:78:in `matches?'
```

Not sure if the following is what you would expect, but as the stacktrace shows `AliasedNegatedMatcher` determines that the `@base_matcher` (`have_rendered`) doesn't respond to a `does_not_match?` method, and the matching logic gets handled up the inheritance tree by `MatcherDelegator`, who calls `matches?` on the base_matcher (`have_rendered`).

The `avoid_redirect_to` negated matcher produces very similar results to `avoid_rendering_template`. With the test result being:

```ruby 
Expected response to be a <redirect>, but was <500>.
```

Stacktrace:
```bash
#../ruby/2.1.0/gems/actionpack-4.0.13/lib/action_dispatch/testing/assertions/response.rb:27:in `assert_response'
#../ruby/2.1.0/gems/actionpack-4.0.13/lib/action_dispatch/testing/assertions/response.rb:53:in `assert_redirected_to'
#../ruby/2.1.0/gems/rspec-rails-3.3.2/lib/rspec/rails/matchers/redirect_to.rb:15:in `block in matches?'
#../ruby/2.1.0/gems/rspec-expectations-3.3.0/lib/rspec/matchers/built_in/base_matcher.rb:47:in `match_unless_raises'
#../ruby/2.1.0/gems/rspec-rails-3.3.2/lib/rspec/rails/matchers/redirect_to.rb:14:in `matches?'
#../ruby/2.1.0/gems/rspec-expectations-3.3.0/lib/rspec/matchers/matcher_delegator.rb:17:in `method_missing'
#../ruby/2.1.0/gems/rspec-expectations-3.3.0/lib/rspec/matchers/aliased_matcher.rb:25:in `method_missing'
#../ruby/2.1.0/gems/rspec-expectations-3.3.0/lib/rspec/matchers/aliased_matcher.rb:78:in `matches?'
```

Not sure if this sheds any further light?
Thanks.

**Comment by JonRowe** [comment]
Ok so I've tested this myself and I can't replicate it, I've opened a PR showing negated examples working correctly (we already had specs for them) (on Rails 4.2, but the PR will test against all our supported versions of Rails) so if that passes we'll need more information to try to fix this...

**Comment by JonRowe** [comment]
The negated examples are all passing for us on Travis so there has to be something else at play here, whats generating `response` here?

**Comment by mattfreer-mdsol** [comment]
Hi @JonRowe, 

Thanks for confirming that this was not a rspec-rails issue. Your findings highlighted that this was probably a configuration issue with our app. Long story short, I discovered that the `test-unit` gem was also being included in the Gemfile, and that this was the culprit; interfering with these specific matchers when negated. Once this dependency was removed the relevant tests passed.

So looks like this issue can be closed :) 
Really appreciate your efforts investigating this.

Thanks,
Matt
